app/services/foreman/plugin/role_lock.rb

```diff
@@ -0,0 +1,72 @@
+module Foreman
+  class Plugin
+    class RoleLock
+      attr_reader :plugin_id
+
+      def initialize(plugin_id)
+        @plugin_id = plugin_id
+      end
+
+      def register_role(name, permissions, role_registry)
+        User.as_anonymous_admin do
+          Role.transaction do
+            role = process_role name, permissions
+            role_registry.role_ids << role.id
+          end
+        end
+      end
+
+      def process_role(name, permissions)
+        role = Role.find_by :name => name
+        if role && role.origin && !role.permission_diff(permissions).empty?
```

Comment:
`!role.permission_diff(permissions).empty?` would be more readable as `role.permission_diff(permissions).present?` (unless nil is a value that was important to accept in this case)